tests/utils/stl/test/file_parsing.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from collections import namedtuple
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, FrozenSet, List, Optional, Set, Tuple, Union
+import itertools
+import os
+import re
+
+import lit.Test
+
+_envlst_cache = dict()
+_preprocessed_file_cache = dict()
+_expected_result_entry_cache = dict()
+
+
+@dataclass
+class _TmpEnvEntry:
+    tags: Set[str] = field(default_factory=set)
+    env: Dict[str, str] = field(default_factory=dict)
+
+
+@dataclass(frozen=True)
+class EnvEntry:
+    def __init__(self, tmp_env: _TmpEnvEntry):
+        object.__setattr__(self, "_tags", frozenset(tmp_env.tags))
+        object.__setattr__(self, "_env_keys", tuple(tmp_env.env.keys()))
+        object.__setattr__(self, "_env_vals", tuple(tmp_env.env.values()))
+
+    def getEnvVal(self, key: str, default: Optional[str] = None) \
+            -> Optional[str]:
+        if key not in self._env_keys:
+            return default
+
+        # TRANSITION: All of this is to avoid having to install frozendict.
+        # Reconsider this at a future date.
+        return self._env_vals[self._env_keys.index(key)]
+
+    def hasTag(self, tag: str) -> bool:
+        return tag in self._env_entry.tags
+
+    _tags: FrozenSet[str]
+    _env_keys: Tuple[str]
+    _env_vals: Tuple[str]
+
+
+@dataclass
+class _ParseCtx:
+    current: List[_TmpEnvEntry] = field(default_factory=list)
+    result: List[List[_TmpEnvEntry]] = field(default_factory=list)
+
+
+_ENV_REGEX = re.compile(r'((?P<tags>(?:\w+,?)+)\t+)?(?P<env_vars_part>.+$)')

```

Comment:
Is the first part of this regex to handle tab-separated tags? I thought we eliminated all of those?

Reply:
We did, this file was partially created by Charlie back when the tags existed, I'll drop it now.

Reply:
Actually, is there any chance at all we could have use for them in the future? Do any of the products that we may potentially at some point open source use them in their `env.lst`s?

The code isn't adding much complexity so I don't see the harm leaving it.

Reply:
> is there any chance at all we could have use for them in the future? Do any of the products that we may potentially at some point open source use them in their `env.lst`s?

Nothing we plan to open source uses them.